docs/tutorial.md

```diff
@@ -0,0 +1,817 @@
+-    if book_action.perform
+-      redirect_to books_url, notice: 'Book was successfully rented.'
+-    else
+-      redirect_to books_url, alert:  book_action.errors.full_messages.to_sentence
+-    end
+-  end
+```
+
+As the last step, we need to change the `config/routes.rb` to use the `granite`
+to mount the `action#projector` into the defined routes.
+
+File: `config/routes.rb`
+```diff
+Rails.application.routes.draw do
+  root 'books#index'
+
+  devise_for :users
+
+  resources :books do
+-   post :rent
++   granite 'BA/book/rent#inline'
+    post 'return', to: 'books#return_book', as 'return'
+  end
+end
+```
+
+!!! warning
+    As it's a tutorial, your next task is to do the same for `return_book`.
+
+    1. Add `projector :inline` to `BA::Book::Return` class.
+    2. Remove the controller method
+    3. Refactor the `config/routes.rb` declaring the `granite 'action#projector'`
+
+
+## Projector Helpers
+
+You can define useful methods for helping you rendering your view and improving
+the experience with your actions. Now, let's create a `button` function,
+to replace the action links in the current list.
+
+First, we need to have a method in our projector that can render the button if
+the action is performable.
+
+It will render nothing if the current user does not have access or it's an
+anonymous session.
+
+We'll render the action name stricken if the action is not performable with the
+error messages in the title, because if people mouse over, they can see the
+"tooltip" with why it's not possible to execute the action.
+
+```ruby
+class InlineProjector < Granite::Projector
+
+  # ...
+  # The previous methods remain here
+  # ...
+
+  def button(link_options = {})
+    return unless action.allowed?
+    if action.performable?
+      h.link_to action_label, perform_path, method: :post
+    else
+      h.content_tag(:strike, action_label, title: action.errors.full_messages.to_sentence)
+    end
+  end
+
+  def action_label
+    action.class.name.demodulize.underscore.humanize
+  end
+end
+```
+
+And now, we can replace the links with the new `button` function:
+
+```erb
+  <tbody>
+    <% @books.each do |book| %>
+      <tr>
+        <td><%= book.title %></td>
+        <td><%= Ba::Book::Rent.as(current_user).new(book).inline.button%></td>
+        <td><%= Ba::Book::Return.as(current_user).new(book).inline.button%></td>
+        <td>... more links here ...</td>
+      </tr>
+    <% end %>
+  </tbody>
+```
+
+Now it's clear, and the "Return" link will appear only for the user
+that rented the book.
+
+Look a print screen comparing two users looking for two books. The user from
+the left side has rented "Learn to fly" book.
+
+Now, look that does not make sense to have the Rent appearing stricken when
+the person is renting the book. We can change the policies to hide the
+action link.
+
+```diff
+@@ -1,6 +1,7 @@
+ # frozen_string_literal: true
+
+ class Ba::Book::Rent < BaseAction
++  allow_if { !performer.renting?(subject) }
```

Comment:
This operates on the subject and I think should be a precondition.

Reply:
This one was an example to show more complex policies but it was a bad decision. My idea was to show the link disabled or enabled based on the policies but in this way, it will disappear the link. I will just cut down this part and make it a precondition.